ansible_kind/scripts/multiclusterkind/util.sh

```diff
@@ -0,0 +1,296 @@
+
+  local counter=0
+  while ! eval "${condition}" | grep "${msg}"; do
+    if [[ "${counter}" = "0" ]]; then
+      echo -n "${start_msg}"
+    fi
+
+    if [[ -z "${timeout}" || "${counter}" -lt "${timeout}" ]]; then
+      counter=$((counter + 1))
+      if [[ -n "${timeout}" ]]; then
+        echo -n '.'
+      fi
+      sleep 1
+    else
+      echo -e "\n[ERROR] Timeout waiting for condition."
+      return 1
+    fi
+  done
+
+  if [[ "${counter}" != "0" && -n "${timeout}" ]]; then
+    echo ' done'
+  fi
+}
+
+# util::wait_file_exist checks if a file exists, if not, wait until timeout
+function util::wait_file_exist() {
+    local file_path=${1}
+    local timeout=${2}
+    for ((time=0; time<timeout; time++)); do
+        if [[ -e ${file_path} ]]; then
+            return 0
+        fi
+        sleep 1
+    done
+    return 1
+}
+
+
+# util::create_cluster creates a kubernetes cluster
+# util::create_cluster creates a kind cluster and don't wait for control plane node to be ready.
+# Parmeters:
+#  - $1: cluster name, such as "host"
+#  - $2: KUBECONFIG file, such as "/var/run/host.config"
+#  - $3: node docker image to use for booting the cluster, such as "kindest/node:v1.19.1"
+#  - $4: log file path, such as "/tmp/logs/"
+function util::create_cluster() {
+  local cluster_name=${1}
+  local kubeconfig=${2}
+  local kind_image=${3}
+  local log_path=${4}
+  local cluster_config=${5:-}
+
+  mkdir -p "${log_path}"
+  rm -rf "${log_path}/${cluster_name}.log"
+  rm -f "${kubeconfig}"
+  nohup kind delete cluster --name="${cluster_name}" >> "${log_path}"/"${cluster_name}".log 2>&1 && kind create cluster --name "${cluster_name}" --kubeconfig="${kubeconfig}" --image="${kind_image}" --config="${cluster_config}" >> "${log_path}"/"${cluster_name}".log 2>&1 &
+  echo "Creating cluster ${cluster_name}"
+}
+
+# This function returns the IP address of a docker instance
+# Parameters:
+#  - $1: docker instance name
+
+function util::get_docker_native_ipaddress(){
+  local container_name=$1
+  docker inspect --format='{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}' "${container_name}"
+}
+
+# This function returns the IP address and port of a specific docker instance's host IP
+# Parameters:
+#  - $1: docker instance name
+# Note:
+#   Use for getting host IP and port for cluster
+#   "6443/tcp" assumes that API server port is 6443 and protocol is TCP
+
+function util::get_docker_host_ip_port(){
+  local container_name=$1
+  docker inspect --format='{{range $key, $value := index .NetworkSettings.Ports "6443/tcp"}}{{if eq $key 0}}{{$value.HostIp}}:{{$value.HostPort}}{{end}}{{end}}' "${container_name}"
+}
+
+# util::check_clusters_ready checks if a cluster is ready, if not, wait until timeout
+function util::check_clusters_ready() {
+  local kubeconfig_path=${1}
+  local context_name=${2}
+
+  echo "Waiting for kubeconfig file ${kubeconfig_path} and clusters ${context_name} to be ready..."
```

Comment:
@sandeepAarna I tried just the standalone bash scripts in the VM (vishsandbox) on GCP cluster and noted that the fourth cluster is not getting created
![Screen Shot 2022-11-08 at 4 53 53 PM](https://user-images.githubusercontent.com/62150532/200693523-3dd87e44-a8dd-4015-8d05-f9c6eafc694b.png)